scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/SparkeyUri.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2017 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.extra.sparkey
+
+import java.io.File
+import java.net.URI
+import java.nio.ByteBuffer
+import java.nio.file.{Files, Paths}
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.google.common.base.Charsets
+import com.google.common.hash.Hashing
+import com.spotify.scio.util.ScioUtil
+import com.spotify.sparkey.{CompressionType, Sparkey, SparkeyReader}
+import org.apache.beam.sdk.options.{GcsOptions, PipelineOptions}
+import org.apache.beam.sdk.util.GcsUtil
+import org.apache.beam.sdk.util.gcsfs.GcsPath
+
+import scala.util.Try
+
+/**
+ * Represents the base URI for a Sparkey index and log file, either on the local file
+ * system or on GCS. For GCS, the URI should be in the form
+ * 'gs://<bucket>/<path>/<sparkey-prefix>'. For local files, the URI should be in the form
+ * '/<path>/<sparkey-prefix>'. Note that the URI must not be a folder or GCS bucket as the URI is
+ * a base path representing two files - <sparkey-prefix>.spi and <sparkey-prefix>.spl.
+ */
+trait SparkeyUri {
+  val basePath: String
+  def getReader: SparkeyReader
+  def exists: Boolean
+  override def toString: String = basePath
+}
+
+object SparkeyUri {
+  /** Create a [[SparkeyUri]] from a string. */
+  def apply(path: String, opts: PipelineOptions): SparkeyUri =
+    if (ScioUtil.isGcsUri(new URI(path))) {
+      new GcsSparkeyUri(path, opts.as(classOf[GcsOptions]))
+    } else {
+      new LocalSparkeyUri(path)
+    }
+}
+
+private case class LocalSparkeyUri(basePath: String) extends SparkeyUri {
+  override def getReader: SparkeyReader = Sparkey.open(new File(basePath))
+  override def exists: Boolean =
+    (new File(basePath + ".spi").exists || new File(basePath + ".spl").exists)
+}
+
+private case class GcsSparkeyUri(basePath: String,
+                                          @transient options: GcsOptions) extends SparkeyUri {
```

Comment:
Indentation.

It might be clearer if it's a simple `private class GcsSparkeyUri` since we don't need the extra stuff from a case class. That way you pass `options` in as an argument and can get rid of `@transient`.

Same goes for `LocalSparkeyUri`, `private class LocalSparkeyUri(val basePath: String)` is clearer IMO.